python/cucim/src/cucim/core/operations/color/stain_normalizer.py

```diff
@@ -0,0 +1,821 @@
+        above the specified `source_intensity` and then normalize by
+        `source_intensity` so that `image` intensities are <= 1.0. Only used
+        when `image_type=='intensity'`.
+    alpha : float, optional
+        Algorithm parameter controlling the ``[alpha, 100 - alpha]``
+        percentile range used as a robust [min, max] estimate.
+    beta : float, optional
+        Absorbance (optical density) threshold below which to consider pixels
+        as transparent. Transparent pixels are excluded from the estimation.
+
+    Additional Parameters
+    ---------------------
+    channel_axis : int, optional
+        The axis corresponding to color channels (default is the last axis).
+    image_type : {'intensity', 'absorbance'}, optional
+        With the default `image_type` of `'intensity'`, the image will be
+        transformed to `absorbance` units via ``image_to_absorbance``. If
+        the input `image` is already an absorbance image, then `image_type`
+        should be set to `'absorbance'` instead.
+    append_third_column : bool, optional
+        If True, a dummy stain vector, orthogonal to the first two will be
+        added.
+
+    Returns
+    -------
+    stain_coeff : cp.ndarray
+        Stain attenuation coefficient matrix derived from the image, where
+        the first column corresponds to H, the second column is E and the rows
+        are RGB values.
+
+    References
+    ----------
+    .. [1] M. Macenko et al., "A method for normalizing histology slides for
+           quantitative analysis," 2009 IEEE International Symposium on
+           Biomedical Imaging: From Nano to Macro, 2009, pp. 1107-1110,
+           doi: 10.1109/ISBI.2009.5193250.
+           http://wwwx.cs.unc.edu/~mn/sites/default/files/macenko2009.pdf
+    """
+
+    _validate_image(image)
+    channel_axis = _prep_channel_axis(channel_axis, image.ndim)
+    if alpha < 0 or alpha > 100:
+        raise ValueError("alpha must be a percentile in range [0, 100].")
+    if beta < 0:
+        raise ValueError("beta must be nonnegative.")
+
+    # convert to absorbance (optical density) matrix of shape (3, n_pixels)
+    absorbance = _image_to_absorbance_matrix(
+        image,
+        source_intensity=source_intensity,
+        image_type=image_type,
+        channel_axis=channel_axis,
+    )
+
+    # remove transparent pixels
+    absorbance = absorbance[:, cp.all(absorbance > beta, axis=0)]
+    if absorbance.size == 0:
+        raise ValueError(
+            "All pixels of the input image are below the threshold."
+        )
+
+    # compute eigenvectors (do small 3x3 matrix calculations on the host)
+    cov = _covariance(absorbance, rowvar=True)
+    cov = cp.asnumpy(cov).astype(np.float32, copy=False)
+    _, ev = np.linalg.eigh(cov)
+    ev = ev[:, [2, 1]]
+    # flip to ensure positive first coordinate so arctan2 angles are about 0
+    if ev[0, 0] < 0:
+        ev[:, 0] *= -1
+    if ev[0, 1] < 0:
+        ev[:, 1] *= -1
+
```

Comment:
This flipping was added here based on an MIT-licensed implementation in:
https://github.com/Peter554/StainTools/blob/2089900d11173ee5ea7de95d34532932afd3181a/staintools/stain_extraction/macenko_stain_extractor.py#L29-L37